Feature : RMSD restraints

BenTanYK opened the pull request:
# Pull request for RMSD restraints feature

Changes proposed in this pull request:

This PR contains a prototype implementation of an RMSD restraints feature. RMSD restraints are applied in a similar style to the positional/angular/dihedral restraints already implemented in Sire. The restraining potential is applied using the OpenMM CustomCVForce/RMSDForce functionality.

The restraints are implemented as follows: 
* Code to create an RMSD restraints container is added to the C++ layer at corelib/src/libs/SireMM/rmsdrestraints.cpp (+ corresponding header file). This allows the required restraint parameters (reference positions, atom selection, force constant, well width, alchemical parameter) to be stored in an analogous way to the previously implemented Sire restraints.
* The python wrappers are recompiled following https://github.com/OpenBioSim/sire/tree/devel/docker/sire-generate-wrappers#using-the-container-against-an-external-sire-directory .
* RMSD restraints are added to the python API at src/sire/restraints/_restraints.py. The user input is very similar to positional restraints.
* Implementation of the harmonic/flat-bottomed potential to restrain the desired RMSD is provided at wrapper/Convert/SireOpenMM/sire_to_openmm_system.cpp
* The lambda lever code at wrapper/Convert/SireOpenMM/lambdalever.cpp is updated to make RMSD restraints compatible with the Sire alchemical restraint scaling functionality.

Declarations:

* I confirm that I have merged the latest version of `devel` into this branch before issuing this pull request 
* I confirm that I have added a test for any new functionality in this pull request
* I confirm that I have added documentation (a new tutorial page) for the new functionality in this pull request
* I confirm that I have added a changelog entry to the changelog at doc/source/changelog.rst
* I confirm that I have permission to release this code under the GPL3 license

## Suggested reviewers:
@chryswoods, @lohedges @akalpokas @jmichel80 

## Extensive testing

Extensive testing of the RMSD restraint feature for the 1UBQ test system can be found in the public repository https://github.com/BenTanYK/sire_rmsd_restraints.git at https://github.com/BenTanYK/sire_rmsd_restraints/blob/191857cca1b7b498b08182456a4c022683070835/testing.ipynb

## Acknowledgements

Thanks to Audrius Kalpokas (@akalpokas) for his advice and input in developing this feature.

lohedges commented:
Thanks for this. There appear to be merge conflicts with `devel`. Could you resolve these locally then push the changes. If you run into issues I can try to do it myself tomorrow.

lohedges commented on src/sire/CMakeLists.txt:
```diff
@@ -4,8 +4,8 @@
 
-cmake_minimum_required(VERSION 3.3.0 FATAL_ERROR)
-cmake_policy(VERSION 3.3)
+cmake_minimum_required(VERSION 3.5.0 FATAL_ERROR)
+cmake_policy(VERSION 3.5)
```
Is CMake 3.5 needed? We've recently updated to 3.3.0 and above, which required updates and fixes to many of the CMake files. If this isn't strictly necessary, could you leave as before. (We plan on updating further at some point so that we are future proofed.)

lohedges commented on wrapper/Convert/SireOpenMM/sire_to_openmm_system.cpp:
```diff
@@ -400,6 +401,111 @@ void _add_positional_restraints(const SireMM::PositionalRestraints &restraints,
+        // Unique CV name
+        std::string rmsd_unique = "rmsd_" + std::to_string(n_CVForces);
+          
+        // Single parameter naming
+        // std::string rho_unique = "rho"
+        // std::string k_unique = "k";
+        // std::string rmsd_b_unique = "rb";
+        // std::string rmsd_unique = "rmsd";
```
Please remove this if it's no longer needed. I assume this was the initial implementation (single RMD restraint) which was superseded by the multi restraint implementation.

lohedges commented on wrapper/Convert/SireOpenMM/sire_to_openmm_system.cpp:
```diff
@@ -400,6 +401,111 @@ void _add_positional_restraints(const SireMM::PositionalRestraints &restraints,
+        auto *restraintff = new OpenMM::CustomCVForce(energy_expression);
+        restraintff->setName("RMSDRestraintForce");
+
+        // Set unique force name
+        // std::string force_name = "RMSDRestraintForce" + std::to_string(n_CVForces);
+        // restraintff->setName(force_name);
```
Remove this too if no longer needed.

BenTanYK commented:
Hi @lohedges many thanks for checking through this, I have gone ahead and changed the  minimum required cmake version back to 3.30 (I had to change this to compile on my workstation) and have also removed the obsolete commented single restraint code, as per your suggestion.

BenTanYK commented:
I have accepted the changes to  the wrapper files from the devel branch, however I am unsure how to resolve the last conflict. This is currently being caused by adding "rmsd" to the ```__all__ ``` list, which I believe is necessary to make the ```restraints.rmsd()``` function importable? 

```python

__all__ = [
    "angle",
    "positional",
    "bond",
    "dihedral",
    "distance",
    "boresch",
    "get_standard_state_correction",
<<<<<<< feature_rmsd_restraints
    "rmsd",
=======
>>>>>>> devel
]

```

lohedges commented:
Sorry, I didn't mean to merge with `devel`, rather all files should be replaced with those from your branch. I think the merges should be okay, but might end up with duplicate includes. For the `__all__`, you should just take the changes from your branch too, i.e.:

```python
__all__ = [
    "angle",
    "positional",
    "bond",
    "dihedral",
    "distance",
    "boresch",
    "get_standard_state_correction",
    "rmsd",
]
```

lohedges commented:
The CI appears to be failing. On macOS I see the following error during the test phase when the Sire Python package is imported:

```
0x0002): symbol not found in flat namespace '__Z27register_CMAPFunction_classv'
```

This is weird, since the function is registered in the following files:

```
grep -r register_CMAPFunc wrapper
wrapper/MM/_MM.main.cpp:    register_CMAPFunction_class();
wrapper/MM/_MM.main.cpp:    register_CMAPFunctions_class();
wrapper/MM/CMAPFunctions.pypp.hpp:void register_CMAPFunctions_class();
wrapper/MM/CMAPFunction.pypp.hpp:void register_CMAPFunction_class();
wrapper/MM/CMAPFunctions.pypp.cpp:void register_CMAPFunctions_class(){
wrapper/MM/CMAPFunction.pypp.cpp:void register_CMAPFunction_class(){
```

The only one modified by your PR is `wrapper/MM/_MM.main.cpp`, but just with the addition of your new class.

For Linux, the CI fails while building the wrappers themselves. Firstly there is _yet_ another `gemmi` issue:

```
CMake Warning at Convert/SireGemmi/CMakeLists.txt:8 (find_package):
  By not providing "Findgemmi.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "gemmi", but
  CMake did not find one.

  Could not find a package configuration file provided by "gemmi" with any of
  the following names:

    gemmiConfig.cmake
    gemmi-config.cmake

  Add the installation prefix of "gemmi" to CMAKE_PREFIX_PATH or set
  "gemmi_DIR" to a directory containing one of the above files.  If "gemmi"
  provides a separate development package or SDK, be sure it has been
  installed.


CMake Warning at Convert/SireGemmi/CMakeLists.txt:77 (message):
  gemmi not found, so we cannot compile the converter.
```

Later there is another CMAP related issue:

```
025-08-07T14:30:06.3380069Z /home/runner/miniconda3/envs/sire_build/conda-bld/sire_1754573015285/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/MM.dir/_MM.main.cpp.o: in function `init_module__MM()':
2025-08-07T14:30:06.3383943Z _MM.main.cpp:(.text+0x232): undefined reference to `register_CMAPFunction_class()'
2025-08-07T14:30:06.3385865Z /home/runner/miniconda3/envs/sire_build/conda-bld/sire_1754573015285/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: _MM.main.cpp:(.text+0x237): undefined reference to `register_CMAPFunctions_class()'
2025-08-07T14:30:06.3388671Z /home/runner/miniconda3/envs/sire_build/conda-bld/sire_1754573015285/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: _MM.main.cpp:(.text+0x23c): undefined reference to `register_CMAPParameter_class()'
```

Again, I don't see any edits that touch the CMAP related functionality, so am confused as to why this is failing. I'll try to create a clean environment and build for myself tomorrow.

lohedges commented on wrapper/MM/SireMM_registrars.cpp:
```diff
@@ -3,195 +3,179 @@
 
 #include "SireMM_registrars.h"
```
It looks like it might be this file that might be the issue. Could you make sure that it is the same as `devel`, plus your edits, i.e. only changed by adding the following in the appropriate places:

```
    #include "rmsdrestraints.h"
    ObjectRegistry::registerConverterFor< SireMM::RMSDRestraint >();
    ObjectRegistry::registerConverterFor< SireMM::RMSDRestraints >();
```

BenTanYK replied:
This may be the case, I believe I first forked the repo in May and I haven't pulled the latest changes recently. I've modified the SireMM_registars.h file so it's the same as the current devel branch (+adding the three lines need for RMSD restraints) - let's see if this helps!

lohedges commented:
Nope, still not working. I'll sort things out on a different branch tomorrow/Monday and will let you know when I'm done.

lohedges commented:
I've redone the merge with `devel` in #343. Hopefully that works.
